File: day-13-bfs-implementation/python/bfs-path.py
from collections import deque

graph = {
  "you": ["alice", "bob", "claire"],
  "alice": ["peggy"],
  "bob": ["anuj", "peggy"],
  "claire": ["thom", "jonny"],
  "anuj": [],
  "peggy": [],
  "thom": [],
  "jonny": []
}

def is_mango_seller(name):
  return name[-1] == "m"

def bfs_path(graph, start):
  queue = deque([start])
  visited = set()
  parent = {start: None} # Who discovered each node

  while queue:
    current = queue.popleft()
    if current not in visited:
      visited.add(current)
      if is_mango_seller(current):
        # Reconstruct path
        path = []
        while current is not None:
          path.append(current)
          current = parent[current]
        return path[::-1] # Reverse to get start->target
      for neighbour in graph[current]:
        if neighbour not in visited:
          parent.setdefault(neighbour, current)
          queue.append(neighbour)

  return None

print(bfs_path(graph, "you"))

File: day-13-bfs-implementation/js/bfs-path.js
graph = {
  you: ["alice", "bob", "claire"],
  alice: ["peggy"],
  bob: ["anuj", "peggy"],
  claire: ["thom", "jonny"],
  anuj: [],
  peggy: [],
  thom: [],
  jonny: [],
};

function isMangoSeller(name) {
  return name[name.length - 1] === "m";
}

function bfsPath(graph, start) {
  const queue = [start];
  const visited = new Set();
  const parent = { [start]: null };

  while (queue.length > 0) {
    const current = queue.shift();
    if (!visited.has(current)) {
      visited.add(current);
      if (isMangoSeller(current)) {
        const path = [];
        let node = current;
        while (node !== null) {
          path.push(node);
          node = parent[node];
        }
        return path.reverse();
      }
      for (const neighbour of graph[current]) {
        if (!visited.has(neighbour)) {
          if (!(neighbour in parent)) {
            parent[neighbour] = current;
          }
          queue.push(neighbour);
        }
      }
    }
  }
  return null;
}

console.log(bfsPath(graph, "you"));
